src/compositemark/callout.ts

```diff
@@ -0,0 +1,114 @@
+import {Config} from '../config';
+import {Encoding} from '../encoding';
+import * as log from '../log';
+import {GenericMarkDef, isMarkDef, MarkConfig} from '../mark';
+import {GenericUnitSpec, LayerSpec} from '../spec';
+import {Flag, keys} from '../util';
+import {getMarkDefMixins} from './common';
+
+export const CALLOUT: 'callout' = 'callout';
+export type Callout = typeof CALLOUT;
+
+export type CalloutPart = 'line' | 'label';
+
+const CALLOUT_PART_INDEX: Flag<CalloutPart> = {
+  line: 1,
+  label: 1
+};
+
+export const CALLOUT_PARTS = keys(CALLOUT_PART_INDEX);
+
+// TODO: Currently can't use `PartsMixins<CalloutPart>`
+// as the schema generator will fail
+export type CalloutPartsMinxins = {
+  [part in CalloutPart]?: MarkConfig
+};
+
+export interface CalloutConfig extends CalloutPartsMinxins {
+  /**
+   * Angle of callout line.
+   * __Default value: `"45"`
+   */
+  calloutAngle?: number;
+  /**
+   * Offset of callout line.
+   * __Default value: `"0"`
+   */
+  calloutOffset?: number;
+  /**
+   * Length of callout length.
+   * __Default value: `"30"`
+   */
+  calloutLength?: number;
+  /**
+   * Offset of callout label.
+   * __Default value: `"0"`
+   */
+  labelOffset?: number;
+}
+
+export interface CalloutDef extends GenericMarkDef<Callout>, CalloutConfig {}
+
+export interface CalloutConfigMixins {
+  /**
+   * Callout Rule Config
+   * @hide
+   */
+  callout?: CalloutConfig;
+}
+
+
+export function normalizeCallout(spec: GenericUnitSpec<Encoding<string>, Callout | CalloutDef>, config: Config): LayerSpec {
+  // TODO:  determine what's the general rule for applying selection for composite marks
+  const {mark, selection: _sel, projection: _p, encoding, ...outerSpec} = spec;
+  const markDef = isMarkDef(mark) ? mark : {type: mark};
+
+  const calloutAngle: number = markDef.calloutAngle || config.callout.calloutAngle;
```

Comment:
This is probably a better pattern

```
const markDef = {
    ...config.callout
    ...isMarkDef(mark) ? mark : {type: mark}
};

const {calloutAngle, calloutOffset, calloutLength, labelOffset} = markDef;
```


(If this is applicable for boxplot, please submit a PR to refactor too.)